masu/processor/tasks.py

```diff
@@ -101,3 +103,63 @@ def _get_report_files(customer_name,
         return []
 
     return reports
+
+
+@shared_task(name='masu.processor.tasks.process_report_file', queue_name='process')
+def process_report_file(schema_name, report_path, compression):
+    """Shared celery task to process report files asynchronously."""
+    _process_report_file(schema_name, report_path, compression)
+
+
+def _process_report_file(schema_name, report_path, compression):
+    """
+    Task to process a Cost Usage Report.
+
+    Args:
+        schema_name (String) db schema name
+        report_path (String) path to downloaded reports
+        compression (String) 'PLAIN' or 'GZIP'
+
+    Returns:
+        None
+
+    """
+    stmt = ('Processing Report:'
+            ' schema_name: {},'
+            ' report_path: {},'
+            ' compression: {}')
+    log_statement = stmt.format(schema_name,
+                                report_path,
+                                compression)
+    LOG.info(log_statement)
+
+    file_name = report_path.split('/')[-1]
+    stats_recorder = ReportStatsDBAccessor(file_name)
+    cursor_position = stats_recorder.get_cursor_position()
+
+    processor = ReportProcessor(schema_name=schema_name,
+                                report_path=report_path,
+                                compression=compression,
+                                cursor_pos=cursor_position)
+
+    stats_recorder.log_last_started_datetime()
+    last_cursor_position = processor.process()
+    stats_recorder.log_last_completed_datetime()
+    stats_recorder.set_cursor_position(last_cursor_position)
+    stats_recorder.commit()
+
+
+@shared_task(name='masu.processor.tasks.check_report_updates', queue_name='celery')
+def check_report_updates():
+    """Scheduled task to initiate scanning process on a regular interval."""
+    reports = []
+    for account in AccountsAccessor().get_accounts():
```

Comment:
Any reason to do this outside of the Orchestrator?  I was expecting the Orchestrator to be created and then call ```prepare()``` to queue the file download work.  I need to look a little closer but I'm not seeing ```prepare()``` used anywhere in the proposed polling workflow.  It's only in SNS.

Reply:
Import loops prevented me from using Orchestrator.  I tried it about a dozen different ways, with no luck.  It's possible @adberglund can spot what I'm doing wrong (again :sweat_smile: ).

Reply:
Interesting, I'm not following what the import loop is.  Orchestrator would work like this:

1. Init Orchestrator -> Get all accounts ```AccountsAccessor().get_accounts()```
2. orchestrator.prepare() -> walk over all accounts and call ```get_report_files.delay``` 

This is exactly what is happening here, only the logic is being pulled out of the orchestrator.

Reply:
I don't claim to understand it (otherwise I'd fix it, rather than working around it).

All I know is that having Orchestrator in the scheduled task constantly gives `could not import Orchestrator` errors. So, there's something in the chain that isn't quite right.

Reply:
Yeah I understand.  Ironically, your comments around DRY from my last PR really apply here.  We will have 3 ways to kick off the download/process pipeline:

1. Download Endpoint
2. SNS Notification Endpoint
3. Celery polling task

All three of these should create an orchestrator and call ```prepare()``` to queue the download task.  The download endpoint and the celery polling task should do exactly the same thing.  They will create the orchestrator without any parameters to make sure it's getting the full list of accounts.  The SNS notification endpoint is the exception.  That takes a specific bucket and the orchestrator will behave identically as the other case only it has a single account on the list.

This flow is a critical path for masu and it's important that all three paths take exactly the same code path from orchestrator init-time onward.

Lets work through the issue with how the tasks are setup and get this straightened out.  Have you tried just making a new task that runs every X minutes/hours and it does only this:

```
@task_periodic(...)
def poll_for_reports():
    Orchestrator().prepare()
```

Reply:
The import loop is essentially a result of moving all tasks to the same file. The orchestrator is calling one of the tasks, so importing the orchestrator in this file is going to fail. I do agree that if we have an orchestrator as our point of entry, we should use it. The benefit I see to that is that it's a single testable thing, and then we can just call that thing from a few different places (the scheduled polling task, and from the SNS HTTP endpoint) and we have the contract in place that the unit we're calling into is tested. 

To avoid a workaround, we either need separate files for all tasks, or just to move the periodic task to a tasks.py file elsewhere. We could define the periodic task in the celery subpackage? Since it effectively becomes the only direct way we call any of the tasks defined in the processor subpackage, I kind of like it living with the rest of the celery stuff. I believe that would allow importing the orchestrator.

Reply:
@adberglund I was having import issues before moving the code around. So, I don't think it's that.

Reply:
@dccurtis - Well, given that my DRY comments from your last PR _didn't_ block the PR, how about we use the same standard?  This code works, even if it isn't perfect and does carry some tech debt.  We can iterate on it.

Reply:
Sorry, but comments/concerns on the structure of the actual code is different from test code (you still never approved it btw, so..).  We need to be careful about what kind of debt we are willing to take even though we are just working on a prototype.

For me, I want to get the tasks straightened out rather then working around the problem.  If someone else wants to approve and we can fix it later, I guess I can live with that.  IMHO this isn't a good trade off for just getting it in, it's central to the design of masu.